## Context

Starting in 2008, every year Forbes Magazine publishes a list of America's best colleges. When it comes to the question everyone seems to be asking, “Is college worth it? this published list of colleges comes handy to take a decision based on student's requirement or desire.

The mission of the college ranking by Forbes Magazine is to conduct an annual review of the undergraduate institutions that deliver the top academics, best experiences, career success and lowest debt. Whether a school is in the Top 10 or near the bottom of the list, the 650 colleges are the best in the country.

For most families, choosing a four-year college is one of the biggest and most expensive decisions they can make. For students, this time of their life may layout their future plans. So choose carefully.

## About Data

The data set contains the rankings of 650 Unites States colleges along with various other statistics pertaining to each school.

<img src ="https://i.ibb.co/sKwQjfw/Stanford.jpg" width=80% height =40%></img>


### Load required packages

In [1]:
import numpy as np
import pandas as pd

### Load the data into pandas dataframe and view the details

In [2]:
df_college = pd.read_csv('ForbesAmericasTopColleges2019.csv')

### Dimension of the data

*'shape'* is a property associated with DataFrame. It returns a tuple representing the dimensionality of the DataFrame. First value is giving number of rows and second value is giving number of columns.

In [3]:
df_college.shape

(650, 17)

In [4]:
print("The number of rows are: ",df_college.shape[0],"\nThe number of columns are: ",df_college.shape[1])

The number of rows are:  650 
The number of columns are:  17


### Datatype of the fields in the data

'info()' method prints information about a DataFrame including the index dtype and column dtypes, non-null values and memory usage

In [5]:
df_college.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 650 entries, 0 to 649
Data columns (total 17 columns):
Rank                        650 non-null int64
Name                        650 non-null object
City                        650 non-null object
State                       650 non-null object
Public/Private              650 non-null object
Undergraduate Population    650 non-null int64
Student Population          650 non-null int64
Net Price                   648 non-null float64
Average Grant Aid           646 non-null float64
Total Annual Cost           650 non-null int64
Alumni Salary               635 non-null float64
Acceptance Rate             648 non-null float64
SAT Lower                   551 non-null float64
SAT Upper                   551 non-null float64
ACT Lower                   553 non-null float64
ACT Upper                   553 non-null float64
Website                     645 non-null object
dtypes: float64(8), int64(4), object(5)
memory usage: 86.4+ KB


Observations:

We have 5 fields with 'object' datatype, 8 fields of datatype 'float' and 4 fields with 'integer' datatype


### Check the missing values in the data
'isnull()' method detects missing values. It returns a boolean same-sized object indicating if the values are NA. NA values, gets mapped to True values.Everything else gets mapped to False values.

'sum()' method followed by 'isnull()' adds up all the True instances per column to give a summary of missing values per column

In [6]:
df_college.isnull().sum()

Rank                         0
Name                         0
City                         0
State                        0
Public/Private               0
Undergraduate Population     0
Student Population           0
Net Price                    2
Average Grant Aid            4
Total Annual Cost            0
Alumni Salary               15
Acceptance Rate              2
SAT Lower                   99
SAT Upper                   99
ACT Lower                   97
ACT Upper                   97
Website                      5
dtype: int64

Observations:

We observe many missing values in SAT and ACT score fields. 

Let's drop 'SAT' and 'ACT' score columns for our analysis.

Also, there are missing values observed in 'Net Price', 'Average Grant Aid', 'Alumni Salary' and 'Acceptance Rate'. 

Let's drop the records/rows with missing values before proceeding with the analysis

'Website' is unique identifier of each college and is not required for further analysis. So, we can drop this column as well.



### Data Cleanup

Let's first drop all columns that are not need. 

'drop()' method with 'axis=1' indicates that columns have to be dropped. List of column names to be dropped is passed to the method. 'inplace = True' ensures that the columns are dropped from the DataFrame permanently

In [7]:
df_college.drop(['Website', 'SAT Lower','SAT Upper','ACT Lower','ACT Upper'], axis=1, inplace=True)

'dropna()' method drops all rows with NA values (by default axis = 0)

In [8]:
df_college.dropna(inplace=True)

In [9]:
df_college.shape

(628, 12)

Note:

After dropping the rows which had missing values we are left with 628 rows

Also as we have dropped rows with missing values the index is no more continuous and has to be reset

### Reset the index of the dataframe

'reset_index()' resets the index of the DataFrame

'drop' as True, will not try to insert index into DataFrame columns

'inplace' as True resets the index permanently

In [10]:
df_college.reset_index(drop=True, inplace=True)

### Check the first 5 rows of the dataset
'head()' function is used to view the top records. By default it displays 5 rows.

In [11]:
df_college.head()

Rank                                   Name       City State  \
0     1                     Harvard University  Cambridge    MA   
1     2                    Stanford University   Stanford    CA   
2     3                        Yale University  New Haven    CT   
3     4  Massachusetts Institute of Technology  Cambridge    MA   
4     5                   Princeton University  Princeton    NJ   

  Public/Private  Undergraduate Population  Student Population  Net Price  \
0        Private                     13844               31120    14327.0   
1        Private                      8402               17534    13261.0   
2        Private                      6483               12974    18627.0   
3        Private                      4680               11466    20771.0   
4        Private                      5659                8273     9327.0   

   Average Grant Aid  Total Annual Cost  Alumni Salary  Acceptance Rate  
0            49870.0              69600       146800.0              5.0  
1            50134.0              69109       145200.0              5.0  
2            50897.0              71290       138300.0              7.0  
3            43248.0              67430       155200.0              7.0  
4            48088.0              66150       139400.0              6.0

### Display the last five rows of the dataset
'tail()' function is used to view the last records. By default it displays 5 rows.

In [12]:
df_college.tail()

Rank                         Name         City State Public/Private  \
623   645            Morehouse College      Atlanta    GA        Private   
624   646  New Mexico State University   Las Cruces    NM         Public   
625   647     Indiana State University  Terre Haute    IN         Public   
626   648    Emory &amp; Henry College        Emory    VA        Private   
627   650              Catawba College    Salisbury    NC        Private   

     Undergraduate Population  Student Population  Net Price  \
623                      2202                2219    34928.0   
624                     13379               14432     8625.0   
625                     13626               13763    13012.0   
626                      1094                1226    19340.0   
627                      1336                1463    19807.0   

     Average Grant Aid  Total Annual Cost  Alumni Salary  Acceptance Rate  
623            17318.0              48723        98700.0             74.0  
624             9582.0              34720        96700.0             64.0  
625             9297.0              32938        85600.0             85.0  
626            27155.0              48100        70700.0             72.0  
627            23662.0              44921        78100.0             42.0

Observations:

From the head() and tail() output we now see that the indexes are realigned i.e. from 0 through 627 

### Recheck the dimension and info of the data 

In [13]:
print("The number of rows are: ",df_college.shape[0],"\nThe number of columns are: ",df_college.shape[1])

The number of rows are:  628 
The number of columns are:  12


In [14]:
df_college.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 628 entries, 0 to 627
Data columns (total 12 columns):
Rank                        628 non-null int64
Name                        628 non-null object
City                        628 non-null object
State                       628 non-null object
Public/Private              628 non-null object
Undergraduate Population    628 non-null int64
Student Population          628 non-null int64
Net Price                   628 non-null float64
Average Grant Aid           628 non-null float64
Total Annual Cost           628 non-null int64
Alumni Salary               628 non-null float64
Acceptance Rate             628 non-null float64
dtypes: float64(4), int64(4), object(4)
memory usage: 59.0+ KB


Note:

The data is now clean and has no missing values

# Data Analysis

### Get Summary of Data

'describe()' method generates descriptive statistics. Descriptive statistics include those that summarize the central
tendency, dispersion and shape of a dataset's distribution, excluding 'NaN' values

In [15]:
df_college.describe()

Rank  Undergraduate Population  Student Population     Net Price  \
count  628.000000                628.000000          628.000000    628.000000   
mean   322.122611              10235.472930        12312.912420  22481.570064   
std    185.360460              11228.048723        13246.660229   8099.771637   
min      1.000000                323.000000          406.000000   1298.000000   
25%    161.750000               2062.750000         2294.500000  16476.000000   
50%    321.500000               4825.500000         6538.500000  22065.000000   
75%    479.250000              15867.000000        18357.250000  27880.750000   
max    650.000000              65100.000000        75044.000000  47270.000000   

       Average Grant Aid  Total Annual Cost  Alumni Salary  Acceptance Rate  
count         628.000000         628.000000     628.000000       628.000000  
mean        19971.522293       50722.579618   98616.242038        61.721338  
std         11214.740998       12670.126944   14026.078146        21.795857  
min          2975.000000        8646.000000   70700.000000         5.000000  
25%          9253.000000       40076.000000   88600.000000        49.000000  
50%         19605.000000       50478.000000   96300.000000        67.000000  
75%         27431.000000       60965.000000  105400.000000        78.000000  
max         50897.000000       75735.000000  158200.000000       100.000000

### 1) Which are the top 5 colleges listed in the Forbes Magazine 2019?

<img src ="https://i.ibb.co/dj0PFhK/Harvard.jpg" width=80% height =40%></img>


As we need the 'Name' of the colleges, we first pick the 'Name' column and then display the top 5 rows using head() function

In [16]:
df_college.Name.head() 

0                       Harvard University
1                      Stanford University
2                          Yale University
3    Massachusetts Institute of Technology
4                     Princeton University
Name: Name, dtype: object

We can also save the information in an object 'top5' and display it

In [17]:
top5 = df_college.Name.head() 
top5

0                       Harvard University
1                      Stanford University
2                          Yale University
3    Massachusetts Institute of Technology
4                     Princeton University
Name: Name, dtype: object

Note:

Since the data is ordered as per the rank of the universities we can just pick up the top 5 colleges.

### 2) Which state and city do the top 5 colleges belong to?

As we need the 'Name', 'City' and 'State' of the colleges, we first pick the required columns and then display the top 5 rows using head() function

In [18]:
df_college[['Name','City','State']].head(5)

Name       City State
0                     Harvard University  Cambridge    MA
1                    Stanford University   Stanford    CA
2                        Yale University  New Haven    CT
3  Massachusetts Institute of Technology  Cambridge    MA
4                   Princeton University  Princeton    NJ

In [19]:
df_college.iloc[0:5,0:4]

Rank                                   Name       City State
0     1                     Harvard University  Cambridge    MA
1     2                    Stanford University   Stanford    CA
2     3                        Yale University  New Haven    CT
3     4  Massachusetts Institute of Technology  Cambridge    MA
4     5                   Princeton University  Princeton    NJ

Note:

As 'City' and 'State' are immediate columns after 'Name', iloc can also be used to get the results

### 3) How many states and which all states have the colleges listed by Forbes?

'nunique()' returns number of unique elements in the object. As we want number of unique states, we first pick up the 'State' column of the DataFrame and then call nunique() function

In [20]:
df_college.State.nunique()

51

'unique()' function returns unique values of Series object. Hence, we first pick up the 'State' column of the DataFrame and then call unique() function to get all all the States

In [21]:
df_college.State.unique()

array(['MA', 'CA', 'CT', 'NJ', 'PA', 'RI', 'NC', 'NH', 'NY', 'DC', 'IL',
       'IN', 'MI', 'TX', 'MD', 'ME', 'TN', 'MO', 'VA', 'VT', 'MN', 'GA',
       'WA', 'WI', 'FL', 'OH', 'IA', 'CO', 'UT', 'OR', 'LA', 'SC', 'DE',
       'AL', 'AZ', 'OK', 'NE', 'KY', 'KS', 'AR', 'WY', 'NV', 'MS', 'ND',
       'HI', 'MT', 'SD', 'NM', 'ID', 'WV', 'AK'], dtype=object)

### 4) In all how many private and public colleges are there in the given list?

'Public/Private' column is of object data type in the DataFrame.

'value_count()' returns a series containing counts of unique values

In [22]:
df_college['Public/Private'].value_counts()

Private    384
Public     244
Name: Public/Private, dtype: int64

Hence, there are 384 private and 244 public colleges in the given list

### 5) From the list how many colleges are there within each State? Which State has highest number of colleges?

As we have to find number of colleges within each State, we first group (groupby()) the data based on 'State'. For each State we then count the colleges using number of unique values (nunique())  in 'Name' field.

As we want the 'State' with highest number of colleges, we sort the values (sort_values()) in descending order to get the results.

<img src ="https://i.ibb.co/GcnXVC0/New-York.jpg" width=80% height =40%></img>

In [23]:
df_college.groupby(by='State').Name.nunique().sort_values(ascending = False)

State
NY    64
CA    56
PA    39
MA    31
OH    29
IL    25
TX    25
FL    20
NC    19
VA    18
IN    18
MN    18
MI    16
GA    14
TN    14
NJ    13
IA    13
WI    13
WA    12
KY    12
MD    12
MO    12
CO    11
OR    11
CT     9
LA     8
SC     8
AL     7
RI     7
NE     6
OK     6
MS     5
DC     5
ME     5
AZ     4
MT     4
ID     4
SD     4
UT     4
VT     4
AR     4
KS     3
NH     3
NM     3
AK     2
ND     2
NV     2
DE     1
HI     1
WV     1
WY     1
Name: Name, dtype: int64

### 6) What is percentage of undergraduates in each college? (Add a new field to calculate)

Divsion of 'Undergraduate Population' by 'Student Population' multiplied by 100 will give percentage of undergraduates.

round() method will round off the derived value to 2 decimal places which is then stored in new field 'Percentage UG'

In [24]:
df_college['Percentage UG'] = round(df_college['Undergraduate Population']/df_college['Student Population']*100,2)

In [25]:
df_college.head()

Rank                                   Name       City State  \
0     1                     Harvard University  Cambridge    MA   
1     2                    Stanford University   Stanford    CA   
2     3                        Yale University  New Haven    CT   
3     4  Massachusetts Institute of Technology  Cambridge    MA   
4     5                   Princeton University  Princeton    NJ   

  Public/Private  Undergraduate Population  Student Population  Net Price  \
0        Private                     13844               31120    14327.0   
1        Private                      8402               17534    13261.0   
2        Private                      6483               12974    18627.0   
3        Private                      4680               11466    20771.0   
4        Private                      5659                8273     9327.0   

   Average Grant Aid  Total Annual Cost  Alumni Salary  Acceptance Rate  \
0            49870.0              69600       146800.0              5.0   
1            50134.0              69109       145200.0              5.0   
2            50897.0              71290       138300.0              7.0   
3            43248.0              67430       155200.0              7.0   
4            48088.0              66150       139400.0              6.0   

   Percentage UG  
0          44.49  
1          47.92  
2          49.97  
3          40.82  
4          68.40

In [ ]:
df[['Name','Percentage UG']].head()

Note:

New field gets added at end of the DataFrame

### 7) Which are the 5 colleges with highest percentage of undergraduates?

Let's just pick up 'Name' and 'Percentage UG' field, sort by 'Percentage UG' in descending order to get the desired ouput

In [26]:
df_college[['Name','Percentage UG']].sort_values(by='Percentage UG', ascending = False).head(5)

Name  Percentage UG
556  University of Texas, Rio Grande Valley          99.99
246            Saint John's University (MN)          99.94
552                      Oakland University          99.94
357                    Salisbury University          99.92
338                     St. Norbert College          99.86

### 8) 5 colleges that receive highest Average Grant Aid?

Sort 'Average Grant Aid' in descending order and pick up the top 5 colleges using 'Name' field

<img src ="https://i.ibb.co/TMtRJ54/Yale.jpg" width=80% height =40%></img>

In [38]:
df_college.sort_values(by='Average Grant Aid', ascending = False).Name.head(5)

2         Yale University
1     Stanford University
0      Harvard University
26        Amherst College
13    Columbia University
Name: Name, dtype: object


### 9) What is the average Undergraduate population in State of 'CA'?

First lets filter the DataFrame and get only rows with 'State' as 'CA'

We then pick up the 'Undergraduate Population' for the above subset and calculate its mean

round off the final value to 2 decimal places using the round() function

In [28]:
round(df_college[df_college.State == 'CA']['Undergraduate Population'].mean(),2)

13186.91

### 10) Which are the colleges where chances of admission is high?

As per the data dictionary 'Acceptance Rate' field provides information on chances of admission. We can sort this column in descending order to get the results

In [29]:
df_college.sort_values(by='Acceptance Rate', ascending = False).head(5)

Rank                                Name         City State  \
588   606        University of Texas, El Paso      El Paso    TX   
607   627  University of Southern Mississippi  Hattiesburg    MS   
339   347               University of Wyoming      Laramie    WY   
616   638             Wright State University       Dayton    OH   
300   308              University of Kentucky    Lexington    KY   

    Public/Private  Undergraduate Population  Student Population  Net Price  \
588         Public                     24386               25078     8135.0   
607         Public                     13606               14478    13178.0   
339         Public                     10874               12397    13070.0   
616         Public                     14332               15957    16249.0   
300         Public                     23718               29465    18527.0   

     Average Grant Aid  Total Annual Cost  Alumni Salary  Acceptance Rate  \
588             9183.0              36955        92100.0            100.0   
607             9292.0              24883        83900.0             98.0   
339             7422.0              31387        98800.0             97.0   
616             6546.0              32506        87800.0             97.0   
300             9387.0              45634        96400.0             96.0   

     Percentage UG  
588          97.24  
607          93.98  
339          87.71  
616          89.82  
300          80.50

### 11) Find if there are any 'Art' colleges in the list

To find a specific string value 'Art' we will use the 'contains()' method. We will pass the string in the paranthesis which needs to be located in the dataframe variable.

In [30]:
df_college[df_college['Name'].str.contains("Art")]

Rank                               Name           City State  \
295   303  Maryland Institute College of Art      Baltimore    MD   
393   402   California Institute of the Arts  Santa Clarita    CA   

    Public/Private  Undergraduate Population  Student Population  Net Price  \
295        Private                      2041                2128    38763.0   
393        Private                       991                1498    46101.0   

     Average Grant Aid  Total Annual Cost  Alumni Salary  Acceptance Rate  \
295            18939.0              62850        84500.0             62.0   
393            15401.0              68993        91000.0             24.0   

     Percentage UG  
295          95.91  
393          66.15

### 12) Lets divide the Alumni Salary equally into 3 buckets (Low Salary - L; Medium Salary - M; High Salary - H ) and check which colleges falls in 'H' bucket

In [31]:
df_college_new = df_college.copy()

Take a copy of the original DataFrame to work on

In [32]:
df_college_new['Alumni Sal Level']=pd.cut(df_college_new['Alumni Salary'], 3, labels =["L", "M", "H"])

'cut()' function bins values into discrete intervals. 'cut()' function is used when you need to segment and sort data values into bins. This function is also useful for going from a continuous variable to a categorical variable.

In the above code we have created 3 bins using 'Alumni Salary' column. Lowest range of salary is labelled as 'L', middle range is labelled as 'M' and highest range of salary is labelled as 'H'. These labels are saved in a new column 'Alumni Sal Level'.

Note: The entire range of the 'Alumni Salary' has been divided into 3 equal ranges i.e. 3 bins

In [33]:
df_college_new.head()

Rank                                   Name       City State  \
0     1                     Harvard University  Cambridge    MA   
1     2                    Stanford University   Stanford    CA   
2     3                        Yale University  New Haven    CT   
3     4  Massachusetts Institute of Technology  Cambridge    MA   
4     5                   Princeton University  Princeton    NJ   

  Public/Private  Undergraduate Population  Student Population  Net Price  \
0        Private                     13844               31120    14327.0   
1        Private                      8402               17534    13261.0   
2        Private                      6483               12974    18627.0   
3        Private                      4680               11466    20771.0   
4        Private                      5659                8273     9327.0   

   Average Grant Aid  Total Annual Cost  Alumni Salary  Acceptance Rate  \
0            49870.0              69600       146800.0              5.0   
1            50134.0              69109       145200.0              5.0   
2            50897.0              71290       138300.0              7.0   
3            43248.0              67430       155200.0              7.0   
4            48088.0              66150       139400.0              6.0   

   Percentage UG Alumni Sal Level  
0          44.49                H  
1          47.92                H  
2          49.97                H  
3          40.82                H  
4          68.40                H

In [34]:
df_college_new['Alumni Sal Level'].value_counts()

L    383
M    217
H     28
Name: Alumni Sal Level, dtype: int64

Note:
There are only 28 colleges in the High Alumni Salary Range

### 13) Let's check the mean Alumni Salary at each level

In [35]:
round(df_college_new.groupby('Alumni Sal Level')['Alumni Salary'].mean(),2)

Alumni Sal Level
L     89835.77
M    109086.64
H    137575.00
Name: Alumni Salary, dtype: float64

### 14) Let's create a new dataframe to further analyze the 28 colleges with High Alumni Salary 

In [36]:
df_high_alumnisal = df_college_new[df_college_new['Alumni Sal Level'] == 'H']

Passing 'include = 'all'' to describe() method provides descriptive summary of all fields in the DataFrame. 'T' property transposes the table which helps in reviewing the statistics per column

In [37]:
df_high_alumnisal.describe(include='all').T

count unique                              top freq  \
Rank                        28    NaN                              NaN  NaN   
Name                        28     28  Stevens Institute of Technology    1   
City                        28     27                        Cambridge    2   
State                       28     12                               CA    6   
Public/Private              28      2                          Private   22   
Undergraduate Population    28    NaN                              NaN  NaN   
Student Population          28    NaN                              NaN  NaN   
Net Price                   28    NaN                              NaN  NaN   
Average Grant Aid           28    NaN                              NaN  NaN   
Total Annual Cost           28    NaN                              NaN  NaN   
Alumni Salary               28    NaN                              NaN  NaN   
Acceptance Rate             28    NaN                              NaN  NaN   
Percentage UG               28    NaN                              NaN  NaN   
Alumni Sal Level            28      1                                H   28   

                             mean      std     min      25%      50%      75%  \
Rank                      68.8571  87.5001       1     8.75     41.5    93.25   
Name                          NaN      NaN     NaN      NaN      NaN      NaN   
City                          NaN      NaN     NaN      NaN      NaN      NaN   
State                         NaN      NaN     NaN      NaN      NaN      NaN   
Public/Private                NaN      NaN     NaN      NaN      NaN      NaN   
Undergraduate Population  6389.54  6290.91     847     2759     4752  7001.75   
Student Population        10603.3  10225.9     861     3253   7019.5  13197.8   
Net Price                 25288.6  9526.08    6758  17913.8  24977.5    34275   
Average Grant Aid         32340.5  14681.9    3597    20604  36200.5  44877.8   
Total Annual Cost         62952.5  14203.5    8646    65024    67931  69918.5   
Alumni Salary              137575  7667.57  129500   132025   135100   140100   
Acceptance Rate           28.0357  21.4898       5     9.75       22       45   
Percentage UG             68.7039  19.0947   40.82    50.03   69.785  80.0725   
Alumni Sal Level              NaN      NaN     NaN      NaN      NaN      NaN   

                             max  
Rank                         333  
Name                         NaN  
City                         NaN  
State                        NaN  
Public/Private               NaN  
Undergraduate Population   31133  
Student Population         41891  
Net Price                  40376  
Average Grant Aid          50897  
Total Annual Cost          74428  
Alumni Salary             158200  
Acceptance Rate               69  
Percentage UG               99.6  
Alumni Sal Level             NaN

Observations:

Out of the 28 shortlisted colleges 22 are private colleges

'CA' is the State which has higest number of these colleges

'Net Price' of these colleges ranges from minimum of 6758 to maximum of 40376

Average 'Acceptance Rate' of these colleges is 28.04

Further analysis can be done on this dataset using the different methods(groupby, sort, nunique) we applied on the whole DataFrame

### Happy Learning!!!